File: src/main/java/com/kingg/api_vacunas_panama/web/dto/UsuarioDto.java
package com.kingg.api_vacunas_panama.web.dto;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.kingg.api_vacunas_panama.persistence.entity.Usuario;
import jakarta.annotation.Nullable;
import jakarta.validation.Valid;
import jakarta.validation.constraints.NotEmpty;
import jakarta.validation.constraints.Pattern;
import jakarta.validation.constraints.Size;

import java.io.Serializable;
import java.time.LocalDateTime;
import java.util.Set;
import java.util.UUID;

/**
 * DTO for {@link Usuario}
 */
public record UsuarioDto(
        UUID id,
        String username,
        @JsonProperty(access = JsonProperty.Access.WRITE_ONLY)
        @Size(min = 8, max = 70, message = "La contraseña no es válida")
        String password,
        @JsonProperty(value = "created_at")
        LocalDateTime createdAt,
        @JsonProperty(value = "updated_at")
        LocalDateTime updatedAt,
        @JsonProperty(value = "last_used")
        LocalDateTime lastUsed,
        @NotEmpty(message = "Los roles no puede estar vacíos")
        @Valid
        Set<RolDto> roles,
        @Nullable
        @Size(max = 15)
        @Pattern(regexp = "^(PE|E|N|[23456789](?:AV|PI)?|1[0123]?(?:AV|PI)?)-(\\d{1,4})-(\\d{1,6})$",
                flags = {Pattern.Flag.CASE_INSENSITIVE, Pattern.Flag.MULTILINE},
                message = "El formato de la cédula no es válido")
        @JsonProperty(access = JsonProperty.Access.WRITE_ONLY)
        String cedula,
        @Nullable
        @Size(max = 20)
        @Pattern(regexp = "^[A-Z0-9]{5,20}$",
                flags = {Pattern.Flag.CASE_INSENSITIVE, Pattern.Flag.MULTILINE},
                message = "El formato del pasaporte no es válido")
        @JsonProperty(access = JsonProperty.Access.WRITE_ONLY)
        String pasaporte,
        @Nullable
        @Size(max = 50)
        @Pattern(regexp = "^.+/DNFD$",
                flags = {Pattern.Flag.CASE_INSENSITIVE, Pattern.Flag.MULTILINE},
                message = "La licenciaFabricante no es válida")
        @JsonProperty(value = "licencia_fabricante", access = JsonProperty.Access.WRITE_ONLY)
        String licenciaFabricante) implements Serializable {
    @Override
    public String toString() {
        return getClass().getSimpleName() + "(" + "id: " + id +
                ", username: " + username + ", createdAt: " + createdAt + ", updatedAt: " + updatedAt + ", lastUsed: " + lastUsed +
                ", roles: [" + roles.toString() + "]" + ", cedula: " + cedula + ", pasaporte: " + pasaporte +
                ", licenciaFabricante: " + licenciaFabricante + ")";
    }
}

File: src/main/kotlin/io/github/kingg22/api/vacunas/panama/modules/usuario/dto/UsuarioDto.kt
package io.github.kingg22.api.vacunas.panama.modules.usuario.dto

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonProperty
import io.github.kingg22.api.vacunas.panama.modules.usuario.domain.UsuarioModel
import io.github.kingg22.api.vacunas.panama.modules.usuario.entity.Usuario
import io.mcarle.konvert.api.KonvertTo
import io.mcarle.konvert.api.Mapping
import jakarta.validation.Valid
import jakarta.validation.constraints.NotEmpty
import jakarta.validation.constraints.PastOrPresent
import jakarta.validation.constraints.Size
import java.io.Serializable
import java.time.LocalDateTime
import java.time.ZoneOffset.UTC
import java.util.UUID

/** DTO for [io.github.kingg22.api.vacunas.panama.modules.usuario.entity.Usuario] */
@JvmRecord
@KonvertTo(
    Usuario::class,
    mappings = [
        Mapping("usuario", "username"), Mapping("clave", "password"),
        Mapping("fabricante", ignore = true), Mapping("persona", ignore = true),
    ],
)
@KonvertTo(UsuarioModel::class)
data class UsuarioDto(
    val id: UUID? = null,

    val username: String? = null,

    @param:JsonProperty(access = JsonProperty.Access.WRITE_ONLY)
    @field:JsonProperty(access = JsonProperty.Access.WRITE_ONLY)
    @param:Size(min = 8, max = 70, message = "La contraseña no es válida")
    @field:Size(min = 8, max = 70, message = "La contraseña no es válida")
    val password: String,

    @param:JsonProperty(value = "created_at")
    @field:JsonProperty(value = "created_at")
    @field:PastOrPresent
    @param:PastOrPresent
    val createdAt: LocalDateTime = LocalDateTime.now(UTC),

    @param:JsonProperty(value = "updated_at", access = JsonProperty.Access.READ_ONLY)
    @field:JsonProperty(value = "updated_at", access = JsonProperty.Access.READ_ONLY)
    @field:PastOrPresent
    @param:PastOrPresent
    val updatedAt: LocalDateTime? = null,

    @param:JsonProperty(value = "last_used", access = JsonProperty.Access.READ_ONLY)
    @field:JsonProperty(value = "last_used", access = JsonProperty.Access.READ_ONLY)
    val lastUsed: LocalDateTime? = null,

    @param:NotEmpty(message = "Los roles no puede estar vacíos")
    @field:NotEmpty(message = "Los roles no puede estar vacíos")
    @param:Valid
    @field:Valid
    val roles: Set<RolDto>,

    @field:JsonIgnore val disabled: Boolean = true,
) : Serializable {

    override fun toString() = UsuarioDto::class.simpleName +
        "(id: $id, username: $username, createdAt: $createdAt, updatedAt: $updatedAt, lastUsed: $lastUsed, roles: $roles)"
}
